golang/cmd/mqtt-to-postgresql/database.go

```diff
@@ -87,1785 +92,1158 @@ func PQErrorHandling(sqlStatement string, err error) {
+	}
+}
 
-		zap.S().Debugf("Got items from queue", prefix, len(itemArray))
+// GetAssetID gets the assetID from the database
+func GetAssetID(customerID string, location string, assetID string) (DBassetID uint32) {
 
-		err = f(itemArray)
-		if err != nil {
-			zap.S().Errorf("Failed to store items in database", prefix)
-			addMultipleItemsToQueue(prefix, pg, itemArray)
-		}
+	// Get from cache if possible
+	var cacheHit bool
+	DBassetID, cacheHit = internal.GetAssetIDFromCache(customerID, location, assetID)
+	if cacheHit { // data found
+		// zap.S().Debugf("GetAssetID cache hit")
+		return
+	}
+
+	err := statement.SelectIdFromAssetTableByAssetIdAndLocationIdAndCustomerId.QueryRow(assetID, location, customerID).Scan(&DBassetID)
+	if err == sql.ErrNoRows {
+		zap.S().Errorf("No Results Found")
+	} else if err != nil {
+		PQErrorHandling("GetAssetID db.QueryRow()", err)
 	}
+
+	// Store to cache if not yet existing
+	go internal.StoreAssetIDToCache(customerID, location, assetID, DBassetID)
+	zap.S().Debugf("Stored AssetID to cache")
+
+	return
 }
 
-func storeItemsIntoDatabaseRecommendation(itemArray []goque.Item) (err error) {
+// GetProductID gets the productID for a asset and a productName from the database
+func GetProductID(DBassetID uint32, productName string) (productID int32, err error) {
 
-	// Begin transaction
-	txn, err := db.Begin()
+	err = statement.SelectProductIdFromProductTableByAssetIdAndProductName.QueryRow(DBassetID, productName).Scan(&productID)
+	if err == sql.ErrNoRows {
+		zap.S().Errorf("No Results Found", DBassetID, productName)
+	} else if err != nil {
+		PQErrorHandling("GetProductID db.QueryRow()", err)
+	}
 
-	if err != nil {
-		err = PQErrorHandlingTransaction("db.Begin()", err, txn)
-		if err != nil {
-			return
-		}
+	return
+}
+
+// GetComponentID gets the componentID from the database
+func GetComponentID(assetID uint32, componentName string) (componentID int32) {
+
+	err := statement.SelectIdFromComponentTableByAssetIdAndComponentName.QueryRow(assetID, componentName).Scan(&componentID)
+	if err == sql.ErrNoRows {
+		zap.S().Errorf("No Results Found")
+	} else if err != nil {
+		PQErrorHandling("GetComponentID() db.QueryRow()", err)
 	}
 
-	var stmt *sql.Stmt
-	stmt, err = txn.Prepare(`
-		INSERT INTO recommendationTable (timestamp, uid, recommendationType, enabled, recommendationValues, recommendationTextEN, recommendationTextDE, diagnoseTextEN, diagnoseTextDE) 
-		VALUES (to_timestamp($1 / 1000.0),$2,$3,$4,$5,$6,$7,$8,$9) 
-		ON CONFLICT (uid) DO UPDATE 
-		SET timestamp=to_timestamp($1 / 1000.0), uid=$2, recommendationType=$3, enabled=$4, recommendationValues=$5, recommendationTextEN=$6, recommendationTextDE=$7, diagnoseTextEN=$8, diagnoseTextDE=$9;`)
+	return
+}
+
+func GetUniqueProductID(aid string, DBassetID uint32) (uid uint32, err error) {
 
-	if err != nil {
-		PQErrorHandling("Prepare()", err)
-		if err != nil {
-			return
+	uid, cacheHit := internal.GetUniqueProductIDFromCache(aid, DBassetID)
```

Comment:
Right now the `GetUniqueProductID` function contains caching (with the `GetUniqueProductIDFromCache` function). Using caching here however can lead to a different behavior then only using the `SelectUniqueProductIdFromUniqueProductTableByUniqueProductAlternativeIdAndAssetIdOrderedByTimeStampDesc` statement because of the desc order.

The UID cached for one AssedID and aid combination won't be changed. Instead of getting the most recent UID we get the oldest UID for the AssetID+aid combination if we use caching. This is a problem especially while using the `storeItemsIntoDatabaseAddParentToChild` where we use the `GetUniqueProductID` function.

As in  #504 the simplest idea might be getting rid of caching in the `GetUniqueProductID` function.